### PR TITLE
Restructure README and add introduction with mdx tests

### PR DESCRIPTION
**Pull request by polytypic**
This PR restructures the README and adds an introduction to the library with mdx checked examples.

(Note that the benchmarks failing is not related to this PR.)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,35 +1,455 @@
...
+```ocaml
+# let rec move ?(backoff = Backoff.default)
+               source
+               target =
+    match Loc.get source with
+    | [] -> raise Exit
+    | (elem::rest) as source' ->
+      let target' = Loc.get target in
+      let ops = [
+        Op.make_cas source source' rest;
```
Comment (talex5) [inline README.md:175]: The use of `source'` to mean "the old value of `source`" is the opposite to other uses I've seen of `'`, which might be confusing.

Comment (polytypic) [reply README.md:175]: Hmm... How about `source_loc` and `target_loc` for the parameters?

Comment (polytypic) [reply README.md:175]: Or perhaps `old_source` and `old_target`, instead.  I guess that is clearer.

**Review by talex5 — APPROVED** [review]
Looks good to me!

**Pull request merged**